src/callback.rs

```diff
@@ -0,0 +1,273 @@
+            // completely read.
+            Some(Box::pin(Cursor::new(buffer.to_vec())) as ContentHandle)
+        } else {
+            None
+        }
+    }
+}
+
+/// A completion progress report, with a final result
+#[derive(Debug)]
+pub enum Progress<T, E> {
+    Update {
+        /// How many elements have been handled
+        now_completed: usize,
+        /// How many elements are expected to be handled
+        total_items: usize,
+    },
+    Complete(Result<T, E>),
+}
+
+/// Wrapper function for callbacks that accept a Progress message
+///
+/// This function has libclamav's `clcb_progress` function signature
+pub(crate) unsafe extern "C" fn progress(
+    total_items: usize,
+    now_completed: usize,
+    context: *mut c_void,
+) -> cl_error_t {
+    // All errors are handled silently as there is no other means to report errors
+    if let Some(sender) = context
+        .cast::<tokio::sync::mpsc::Sender<Progress<(), EngineError>>>()
+        .as_ref()
+    {
+        let _ = sender.blocking_send(Progress::Update {
+            total_items,
+            now_completed,
+        });
+    }
+
+    // ClamAV doesn't specify any action on this value, so it's hardcoded into
+    // the wrapper
+    cl_error_t::CL_SUCCESS
+}
+
+pub(crate) unsafe extern "C" fn engine_pre_scan(
+    fd: c_int,
+    type_: *const c_char,
+    context: *mut c_void,
+) -> cl_error_t {
+    if let Some(cxt) = context.cast::<ScanCbContext>().as_ref() {
+        let file_type = CStr::from_ptr(type_).to_string_lossy();
+
+        let _ = cxt.sender.blocking_send(ScanEvent::PreScan {
+            file: dup_fd_to_file(fd),
+            file_type: file_type.into(),
+        });
+    }
+
+    cl_error_t::CL_CLEAN
+}
+
+pub(crate) unsafe extern "C" fn engine_post_scan(
+    fd: c_int,
+    result: c_int,
+    virname: *const c_char,
+    context: *mut c_void,
+) -> cl_error_t {
+    if let Some(cxt) = context.cast::<ScanCbContext>().as_ref() {
+        let result = result as isize;
+        let match_name = if virname.is_null() {
+            String::from("<NULL>")
+        } else {
+            CStr::from_ptr(virname).to_string_lossy().into()
+        };
+
+        let _ = cxt.sender.blocking_send(ScanEvent::PostScan {
+            file: dup_fd_to_file(fd),
+            result,
+            match_name,
+        });
+    }
+
+    cl_error_t::CL_CLEAN
+}
+
+pub(crate) unsafe extern "C" fn engine_virus_found(
+    fd: c_int,
+    virname: *const c_char,
+    context: *mut c_void,
+) {
+    if let Some(cxt) = context.cast::<ScanCbContext>().as_ref() {
+        let name = CStr::from_ptr(virname).to_string_lossy().into();
+
+        let _ = cxt.sender.blocking_send(ScanEvent::MatchFound {
+            file: dup_fd_to_file(fd),
+            name,
+        });
+    }
+}
+
+pub(crate) unsafe extern "C" fn engine_file_inspection(
+    // NOTE: this file descriptor is unsafe to use after the callback has
+    // returned, even if dup'd. Hence, it's just ignored.
```

Comment:
I didn't investigate much, particuarly given that the behavior is probably different depending on the platform.  If I had to guess, I'd say that the 64K confusion we saw was related to file system caching, and might have even been specific to my development platform.

Copying the buffer is safer, albeit slower. But it could get faster later if we could refcount the buffers.